executor/index_merge_reader.go

```diff
@@ -0,0 +1,738 @@
+			workCh:         workCh,
+			finished:       e.finished,
+			buildTblReader: e.buildFinalTableReader,
+			keepOrder:      false,
+			handleIdx:      e.handleIdx,
+			tblPlans:       e.tblPlans,
+			memTracker: memory.NewTracker(stringutil.MemoizeStr(func() string { return "TableWorker_" + strconv.Itoa(i) }),
+				e.ctx.GetSessionVars().MemQuotaIndexLookupReader),
+		}
+		ctx1, cancel := context.WithCancel(ctx)
+		go func() {
+			worker.pickAndExecTask(ctx1)
+			cancel()
+			e.tblWorkerWg.Done()
+		}()
+	}
+}
+
+func (e *IndexMergeReaderExecutor) buildFinalTableReader(ctx context.Context, handles []int64) (Executor, error) {
+	tableReaderExec := &TableReaderExecutor{
+		baseExecutor: newBaseExecutor(e.ctx, e.schema, stringutil.MemoizeStr(func() string { return e.id.String() + "_tableReader" })),
+		table:        e.table,
+		dagPB:        e.tableRequest,
+		startTS:      e.startTS,
+		streaming:    e.tableStreaming,
+		feedback:     statistics.NewQueryFeedback(0, nil, 0, false),
+		plans:        e.tblPlans,
+	}
+	tableReader, err := e.dataReaderBuilder.buildTableReaderFromHandles(ctx, tableReaderExec, handles)
+	if err != nil {
+		logutil.Logger(ctx).Error("build table reader from handles failed", zap.Error(err))
+		return nil, err
+	}
+	return tableReader, nil
+}
+
+// Next implements Executor Next interface.
+func (e *IndexMergeReaderExecutor) Next(ctx context.Context, req *chunk.Chunk) error {
+	req.Reset()
+	for {
+		resultTask, err := e.getResultTask()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if resultTask == nil {
+			return nil
+		}
+		for resultTask.cursor < len(resultTask.rows) {
+			req.AppendRow(resultTask.rows[resultTask.cursor])
+			resultTask.cursor++
+			if req.NumRows() >= e.maxChunkSize {
+				return nil
+			}
+		}
+	}
+}
+
+func (e *IndexMergeReaderExecutor) getResultTask() (*lookupTableTask, error) {
+	if e.resultCurr != nil && e.resultCurr.cursor < len(e.resultCurr.rows) {
+		return e.resultCurr, nil
+	}
+	task, ok := <-e.resultCh
+	if !ok {
+		return nil, nil
+	}
+	if err := <-task.doneCh; err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// Release the memory usage of last task before we handle a new task.
+	if e.resultCurr != nil {
+		e.resultCurr.memTracker.Consume(-e.resultCurr.memUsage)
+	}
+	e.resultCurr = task
+	return e.resultCurr, nil
+}
+
+// Close implements Exec Close interface.
+func (e *IndexMergeReaderExecutor) Close() error {
+	if !e.workerStarted || e.finished == nil {
+		return nil
+	}
+	close(e.finished)
+	e.processWokerWg.Wait()
+	e.tblWorkerWg.Wait()
+	e.partialWorkerWg.Wait()
+	e.finished = nil
+	e.workerStarted = false
+	// TODO: how to store e.feedbacks
+	return nil
+}
+
+type indexMergeProcessWorker struct {
+	fetchCh  <-chan *lookupTableTask
+	finished <-chan struct{}
+	workCh   chan<- *lookupTableTask
+	resultCh chan<- *lookupTableTask
+	maps     map[int64]byte
+}
+
+func (w *indexMergeProcessWorker) fetchLoop(ctx context.Context, partialWorkerCount int) {
```

Comment:
Since this method works as a single goroutine, it's better to handle panic at the beginning of it.